legate/driver/launcher.py

```diff
@@ -83,6 +84,15 @@ def __init__(self, config: ConfigProtocol, system: System) -> None:
         self._config = config
         self._system = system
 
+        if config.multi_node.ranks == 1:
+            self.rank_id = "0"
+
+        elif "LEGATE_RANK" in system.env:
+            self.rank_id = system.env["LEGATE_RANK"]
```

Comment:
The legate driver is the one that prepares the call to `bind.sh`. That script hasn't run yet, and therefore `LEGATE_RANK` is not available.

What we could do instead is use a placeholder on the synthesized command line, e.g. `%%RANK%%`, for the different wrappers' output filenames. For example, the command for an invocation wrapped with cprofile might be:

```
mpirun -N 2 ./bind.sh legion_python -m cProfile -o legate_%%RANK%%.py-prof prog.py
```

and an invocation for nvprof would be:

```
mpirun -N 2 ./bind.sh nvprof -o legate_%%RANK%%.nvvp legion_python prog.py
```

Then we would have `bind.sh` find the actual global rank on each invocation, and replace all instances of `%%RANK%%` with that.

Reply:
:facepalm: sometimes you get an idea in your head, and it just obscures some important detail...

This seem ok, I don't love it exactly, but I'd rather manage this substitution in one place.  I wonder if just using the env var directly could work
```
-o "legate_${LEGATE_RANK}"
```
with so many layers it's not 100% obvious when the expansion would happen tho I'm guessing also too soon, but maybe there is a way to introduce one layer of quoting that gets "naturally" stripped off as the arg is passed down a level.

Reply:
If you just do a `subprocess` call, then it looks like the envvars don't get naturally stripped off:

```
~/Desktop> echo $HOME
/Users/mpapadakis
~/Desktop> cat a.py
import subprocess
subprocess.check_call(["./bind.sh", "ls", "$HOME"])
~/Desktop> cat bind.sh
#!/bin/bash
echo "gonna execute $@"
"$@"
~/Desktop> python a.py
gonna execute ls $HOME
ls: $HOME: No such file or directory
Traceback (most recent call last):
  File "a.py", line 2, in <module>
    subprocess.check_call(["./bind.sh", "ls", "$HOME"])
  File "/Users/mpapadakis/opt/miniconda3/envs/legate/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['./bind.sh', 'ls', '$HOME']' returned non-zero exit status 1.
```

This would probably work if we used `shell=True`, but that's generally frowned upon https://docs.python.org/3/library/subprocess.html#security-considerations.

Reply:
@manopapad see https://github.com/nv-legate/legate.core/pull/470/commits/8635b8b0c802a07184db0c98b9f996a2c6eb1578 if this looks good I will update the driver unit tests